src/archive_writer.rs

```diff
@@ -302,45 +351,217 @@ fn write_symbol_table_header<W: Write + Seek>(
         let pos = w.stream_position()?;
         print_bsd_member_header(w, pos, name, now(), 0, 0, 0, size)
     } else if is_aix_big_archive(kind) {
-        print_big_archive_member_header(w, "", now(), 0, 0, 0, size, prev_member_offset, 0)
+        print_big_archive_member_header(
+            w,
+            "",
+            now(),
+            0,
+            0,
+            0,
+            size,
+            prev_member_offset,
+            next_member_offset,
+        )
     } else {
         let name = if is_64bit_kind(kind) { "/SYM64" } else { "" };
         print_gnu_small_member_header(w, name.to_string(), now(), 0, 0, 0, size)
     }
 }
 
+fn compute_headers_size(
+    kind: ArchiveKind,
+    num_members: usize,
+    string_member_size: u64,
+    num_syms: u64,
+    sym_names_size: u64,
+    sym_map: Option<&SymMap>,
+) -> io::Result<u64> {
+    let offset_size = if is_64bit_kind(kind) { 8 } else { 4 };
+    let (symtab_size, _) =
+        compute_symbol_table_size_and_pad(kind, num_syms, offset_size, sym_names_size);
+    let compute_symbol_table_header_size = || -> io::Result<u64> {
+        let mut tmp = Cursor::new(Vec::new());
+        write_symbol_table_header(&mut tmp, kind, symtab_size, 0, 0)?;
+        Ok(tmp.into_inner().len().try_into().unwrap())
+    };
+    let header_size = compute_symbol_table_header_size()?;
+    let mut size = u64::try_from("!<arch>\n".len()).unwrap() + header_size + symtab_size;
+
+    if let Some(sym_map) = sym_map {
+        size += header_size + compute_symbol_map_size_and_pad(num_members, sym_map).0;
+        if !sym_map.ec_map.is_empty() {
+            size += header_size + compute_ec_symbols_size_and_pad(sym_map).0;
+        }
+    }
+
+    Ok(size + string_member_size)
+}
+
+struct SymbolicFile<'a> {
+    data: &'a [u8],
+    parsed: File<'a>,
```

Comment:
ar_archive_writer intentionally doesn't parse object files. When used as part of rustc symbols we need to get symbols from LLVM bitcode files, which only LLVM itself can. This is why the `get_symbols` field is used for getting symbols instead. Rustc overrides it with a function which calls into LLVM to get all symbols.

Reply:
I should be able to avoid parsing in most cases, but I'm not sure what to do about AIX symbol tables: we need to know if the obj is 32 or 64 bit to place the symbols in the correct table, and we need to get the alignment. Should I add new callbacks for both of those?

Reply:
Ok, I've moved everything that was using the `object` crate into a new `object_reader` module and replaced the `get_symbol` function pointer with an `ObjectReader` struct that has function pointers for each of the functions that need to parse object files.